CONTENT WARNING: explicit and multiple references to suicide attemps and suicidal ideation

The following is an analysis regarding attempted suicides and suicidal thoughts. Questions around these topics will be directly referenced and mentioned explicitly. While their potential correlations with other parameters will be analysed and discussed at length, suicide attempts and suicidal ideation as concepts will not be discussed. Deaths by suicide will not be referenced or discussed at all.

Please only continue if it comfortable for you to do so.

In [12]:
import pandas as pd
import statsmodels.api as sm
import numpy as np
from sklearn.cross_validation import train_test_split
import seaborn as sb
from sklearn.linear_model import LogisticRegression
from sklearn.feature_selection import RFE
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
import pickle

2016 results ( https://www.datafiles.samhsa.gov/study-dataset/national-survey-drug-use-and-health-2016-nsduh-2016-ds0001-nid17185 )

In [6]:
df = pd.DataFrame.from_csv('NSDUH_2016_Tab.tsv',sep='\t')

/anaconda3/envs/condasc/lib/python2.7/site-packages/ipykernel_launcher.py:1: FutureWarning: from_csv is deprecated. Please use read_csv(...) instead. Note that some of the default arguments are different, so please refer to the documentation for from_csv when changing your function calls
  """Entry point for launching an IPython kernel.


In [7]:
df

FILEDATE  CIGEVER  CIGOFRSM  CIGWILYR  CIGTRY  CIGYFU  CIGMFU  \
QUESTID2                                                                    
11635143  02/28/2018        1        99        99      16    9999      99   
36845143  02/28/2018        1        99        99      15    9999      99   
35755143  02/28/2018        1        99        99      26    9999      99   
94475143  02/28/2018        2         4         4     991    9991      91   
92675143  02/28/2018        1        99        99       5    9999      99   
59596143  02/28/2018        2        99        99     991    9991      91   
41106143  02/28/2018        2        99        99     991    9991      91   
96416143  02/28/2018        2        99        99     991    9991      91   
73716143  02/28/2018        1        99        99      15    9999      99   
76226143  02/28/2018        2        99        99     991    9991      91   
99926143  02/28/2018        1        99        99      12    9999      99   
39446143  02/28/2018        1        99        99      16    9999      99   
14946143  02/28/2018        2         4         4     991    9991      91   
61056143  02/28/2018        2        99        99     991    9991      91   
55466143  02/28/2018        1        99        99     994    9994      94   
83666143  02/28/2018        2        99        99     991    9991      91   
71866143  02/28/2018        1        99        99      16    9999      99   
60397143  02/28/2018        2         4         4     991    9991      91   
59497143  02/28/2018        1        99        99      22    2016       3   
75227143  02/28/2018        1        99        99      14    9999      99   
68927143  02/28/2018        2         4         4     991    9991      91   
31437143  02/28/2018        2         4         4     991    9991      91   
89637143  02/28/2018        1        99        99      35    9999      99   
21957143  02/28/2018        1        99        99      14    9999      99   
29587143  02/28/2018        2        99        99     991    9991      91   
75987143  02/28/2018        2        99        99     991    9991      91   
43328143  02/28/2018        1        99        99      18    2015       6   
30438143  02/28/2018        1        99        99      18    2015      12   
64038143  02/28/2018        1        99        99      18    9999      99   
63899143  02/28/2018        1        99        99      14    9999      99   
...              ...      ...       ...       ...     ...     ...     ...   
43370730  02/28/2018        2        99        99     991    9991      91   
15791730  02/28/2018        1        99        99      18    9999      99   
91901730  02/28/2018        2        99        99     991    9991      91   
13321730  02/28/2018        2        99        99     991    9991      91   
33271730  02/28/2018        2        99        99     991    9991      91   
94171730  02/28/2018        2         4         4     991    9991      91   
30192730  02/28/2018        2         4         4     991    9991      91   
19002730  02/28/2018        2        99        99     991    9991      91   
15212730  02/28/2018        1        99        99      13    9999      99   
81372730  02/28/2018        2         4         4     991    9991      91   
85972730  02/28/2018        2         4         4     991    9991      91   
47582730  02/28/2018        2        99        99     991    9991      91   
15782730  02/28/2018        2        99        99     991    9991      91   
53982730  02/28/2018        1        99        99      18    9999      99   
77393730  02/28/2018        2         4         4     991    9991      91   
22893730  02/28/2018        1        99        99      14    9999      99   
81993730  02/28/2018        1        99        99      22    9999      99   
39193730  02/28/2018        2        99        99     991    9991      91   
19903730  02/28/2018        2         4         4     991    9991      91   
75443730

In [11]:
print('Num adults w/ suicidal thoughts: '+str(len(df[(df.MHSUITHK==1) | (df.ADWRSTHK==1)])))
print('Num adults w/ suicidal plans: '+str(len(df[(df.MHSUIPLN==1) | (df.ADWRSPLN==1)])))
print('Num adults w/ suicidal attempt: '+str(len(df[(df.MHSUITRY==1) | (df.ADWRSATP==1)])))
print('Num adults w/ thoughts of death: '+str(len(df[(df.ADWRDLOT==1) | (df.ADWRDBTR==1)])))
print('Num adults w/ thoughts of death or suicide: '+str(len(df[(df.ADWRDLOT==1) | (df.ADWRDBTR==1) | (df.MHSUITHK==1) | (df.ADWRSTHK==1)])))

Num adults w/ suicidal thoughts: 3930
Num adults w/ suicidal plans: 1444
Num adults w/ suicidal attempt: 958
Num adults w/ thoughts of death: 4612
Num adults w/ thoughts of death or suicide: 5598


In [22]:
results17 = pickle.load(open('logit_2017data','rb'))

In [27]:
def prepData17(df,use_col,use_param):
    rem_val = [81,83,85,89,91,93,94,97,98,99,981,983,985,989,991,993,994,997,998,999,9981,9983,9985,9989,9991,9993,
               9994,9997,9998,9999,186,187,188,180]

    dfc = df.copy()
    dfc = dfc[(dfc.MHSUITHK==0) | (dfc.MHSUITHK==1)]
    
    dfc.HTINCHE2 = np.floor((dfc.HTINCHE2-55)/5)
    dfc.WTPOUND2 = np.floor((dfc.WTPOUND2-75)/55)
    dfc.BMI2 = np.floor((dfc.BMI2-9.3)/14.83)
    dfc.EDUSCHLGO.replace(11,1,inplace=True)
    dfc.BOOKED.replace(3,1,inplace=True)
    dfc.SNYSELL.replace(1,0,inplace=True)
    dfc.SNYSTOLE.replace(1,0,inplace=True)
    dfc.SNYATTAK.replace(1,0,inplace=True)
    dfc.SNYSELL.replace([2,3,4,5],1,inplace=True)
    dfc.SNYSTOLE.replace([2,3,4,5],1,inplace=True)
    dfc.SNYATTAK.replace([2,3,4,5],1,inplace=True)
    
    y = dfc.MHSUITHK
    use_X = dfc[use_col]
    dum_p = []
    
    for param in use_col:
        if use_X[param].max() == 2:
            use_X[param].replace(2,0,inplace=True)
            X_teu[param].replace(2,0,inplace=True)
        elif use_X[param].max() > 2:
            dum_p.append(param)
    
    use_X = pd.get_dummies(use_X,columns=dum_p) 
    return use_X[use_param],y
    

In [36]:
use_col17 = ['CATAG6','HEALTH','MOVSINPYR2','SEXATRACT','IRMARIT','IREDUHIGHST2','NEWRACE2','WRKSTATWK2','IRPINC3',
           'SNRLGIMP','WTPOUND2','BMI2']
use_param17 = ['CATAG6_2', 'CATAG6_3', 'CATAG6_5', 'CATAG6_6', 'HEALTH_1','HEALTH_2', 'HEALTH_4', 'HEALTH_5',
               'MOVSINPYR2_2','MOVSINPYR2_3', 'SEXATRACT_1', 'SEXATRACT_3', 'IRMARIT_1', 'IRMARIT_2','IREDUHIGHST2_6',
               'IREDUHIGHST2_9', 'NEWRACE2_2','NEWRACE2_5', 'NEWRACE2_7','IRPINC3_5', 'IRPINC3_7', 'SNRLGIMP_1']
X16,y16 = prepData17(df,use_col17,use_param17)

In [37]:
y_pred = np.round(results17.predict(X16))
print('fraction wrong: '+str(np.sum(np.abs(y_pred-y16))/len(y16)))
confusion_matrix(y16, y_pred)

fraction wrong: 0.291016917116


array([[28503, 11423],
       [  894,  1504]])

Type 1 error: 11423 (False positive (suicidal ideation in this case))

Type 2 error: 894 (False negative (did not predict SI))

In [39]:
print(classification_report(y16, y_pred))

             precision    recall  f1-score   support

        0.0       0.97      0.71      0.82     39926
        1.0       0.12      0.63      0.20      2398

avg / total       0.92      0.71      0.79     42324



In [43]:
len(df[df.MHSUITHK==1]),len(df[df.MHSUITHK==0])

(2398, 39926)

Model using 2017 data (used undersampling) does not perform well with 2016 data.

Next step: explore with more data and try other methods for dealing with unbalanced sample